src/ck-ldb/TreeLB.C

```diff
@@ -0,0 +1,537 @@
+  if (num_pes > 1)
+  {
+    int num_subtrees = tb.buildSpanningTree(_pes, _pes + num_pes, mcast_bfactor);
+    for (int i = 0; i < num_subtrees; i++)
+      thisProxy[*tb.begin(i)].multicastIDM(mig_order, tb.subtreeSize(i), tb.begin(i));
+  }
+  migrateObjects(mig_order);
+}
+
+void TreeLB::sendDecisionDown(CkMessage* msg)
+{
+  TreeLBMessage* decision = (TreeLBMessage*)msg;
+  int level = decision->level;
+  std::vector<int>& children = comm_children[level];
+  if (children.size() == 0)
+  {
+    receiveDecision(decision, level);
+  }
+  else
+  {
+    // comm logic is free to split (scatter) the message, or send same msg to every child,
+    // etc.
+    std::vector<TreeLBMessage*> decisions;
+    CkAssert(comm_logic[level] != nullptr);
+    comm_logic[level]->splitDecision(decision, decisions);
+    CkAssert(decisions.size() == children.size() + 1);
+    decisions[0]->level = level;
+    receiveDecision(decisions[0], level);
+    for (int i = 0; i < children.size(); i++)
+    {
+      decisions[i + 1]->level = level;
+      thisProxy[children[i]].sendDecisionDown((CkMessage*)decisions[i + 1]);
+    }
+  }
+}
+
+void TreeLB::receiveDecision(TreeLBMessage* decision, int level)
+{
+  // fprintf(stderr, "[%d] TreeLB::receiveDecision, level=%d\n", CkMyPe(), level);
+
+  // incoming and outgoing are integers. logic objects determine and interpret these
+  // values
+  int& incoming = expected_incoming[level];
+  int& outgoing = expected_outgoing[level];
+  logic[level]->processDecision(decision, incoming, outgoing);
+  // fprintf(stderr, "[%d] level=%d incoming=%d outgoing=%d\n", CkMyPe(), level, incoming,
+  // outgoing);
+  if (incoming == 0 && outgoing == 0)
+  {
+    // no exchange with other subtrees, can do lb for my subtree now
+    loadBalanceSubtree(level);
+  }
+  else
+  {
+    // awaiting load from other subtrees
+    if (outgoing > 0 && level > 0)
+    {
+      // need to pass info on actual chares to other subtree(s) (they will need it sooner
+      // or later since some of my chares will be moving there, but I'm not migrating
+      // chares themselves yet, since their final destination is not yet known) at this
+      // level I might not know about individual chares. if that's the case, I have to
+      // delegate to lower levels to pass this info to the other subtree
+      transferLoadToken(decision, level);
+    }
+
+    // if outgoing > 0 and we are in last level (0), concrete objects are moved to
+    // concrete PEs. for chares whose destination is known, the logic object would have
+    // made call to migrate them inside processDecision (above) NOTE: I might be moving
+    // some objects to other subtrees, might need to wait to receive messages from those
+    // subtrees telling me to which PEs to migrate them
+  }
+  // fprintf(stderr, "[%d] TreeLB::receiveDecision, deleting decision msg\n", CkMyPe());
+  delete (CkMessage*)decision;  // FIXME this won't call the subclass' delete method
```

Comment:
Does this leak memory?

Reply:
No, it just wouldn't call a chain of destructors, which don't even exist in any of the message types as far as I can tell. I'm not sure what @juanjgalvez's concern here was.